src/java.base/share/classes/java/util/stream/Gatherers.java

```diff
@@ -0,0 +1,596 @@
+                        downstream.push(List.copyOf(window));
+                }
+        );
+    }
+
+    /**
+     * Gathers elements into sliding windows, sliding out the most previous
+     * element and sliding in the next element for each subsequent window.
+     * If the stream is empty then no window will be produced. If the size of
+     * the stream is smaller than the window size then only one window will
+     * be emitted, containing all elements.
+     *
+     * <p>Example:
+     * {@snippet lang = java:
+     * // will contain: [[1, 2], [2, 3], [3, 4], [4, 5], [5, 6], [6, 7], [7, 8]]
+     * List<List<Integer>> windows =
+     *     Stream.of(1,2,3,4,5,6,7,8).gather(Gatherers.windowSliding(2)).toList();
+     * }
+     *
+     * @param windowSize the size of the windows
+     * @param <TR> the type of elements the returned gatherer consumes
+     *             and the contents of the windows it produces
+     * @return a new gatherer which groups elements into sliding windows
+     * @throws IllegalArgumentException when windowSize is less than 1
+     */
+    public static <TR> Gatherer<TR, ?, List<TR>> windowSliding(int windowSize) {
+        if (windowSize < 1)
+            throw new IllegalArgumentException("'groupSize' must be greater than zero");
+
+        @SuppressWarnings("serial")
+        class SlidingWindow extends ArrayDeque<TR> {
+            boolean firstWindow = true;
+        }
+        return Gatherer.ofSequential(
+                // Initializer
+                SlidingWindow::new,
+
+                // Integrator
+                Integrator.ofGreedy((window, e, downstream) -> {
+                    window.addLast(e);
+                    if (window.size() < windowSize) {
+                        return true;
+                    } else {
+                        var full = List.copyOf(window);
+                        window.removeFirst();
+                        window.firstWindow = false;
+                        return downstream.push(full);
+                    }
+                }),
+
+                // Finisher
+                (window, downstream) -> {
+                    if(window.firstWindow && !downstream.isRejecting() && !window.isEmpty())
+                        downstream.push(List.copyOf(window));
+                }
+        );
+    }
+
+    /**
+     * An operation which performs an ordered, <i>reduction-like</i>,
+     * transformation for scenarios where no combiner-function can be
+     * implemented, or for reductions which are intrinsically
+     * order-dependent.
+     *
+     * <p>This operation always emits a single resulting element.
+     *
+     * <p>Example:
+     * {@snippet lang = java:
+     * // will contain: Optional[123456789]
+     * Optional<String> numberString =
+     *     Stream.of(1,2,3,4,5,6,7,8,9)
+     *           .gather(
+     *               Gatherers.fold(() -> "", (string, number) -> string + number)
```

Comment:
This particular sample is not great, as it's totally replaceable with reduce() (`map(String::valueOf).reduce(String::concat)`), so users may wonder why using fold instead. I can suggest another sample which I show everywhere:
`anyList.stream().gather(Gatherers.fold(() -> 1, (acc, e) -> acc * 31 + Objects.hashCode(e))).findFirst().get();`
(haven't tested, but the idea is like this)
It computes the hashCode of the list, according to the list specification. Here, the accumulator is inherently non-associative, so if you replace it with `reduce()` and make the stream parallel, it will yield the wrong result. But `Gatherers.fold` will work!

If you don't like it, we can think up something else, but concatenation sample is certainly not good.

Reply:
Perhaps the best thing here is to change such that the supplier supplies something non-zero like "String: "?

Reply:
Such an example still looks artificial for me, as it's easy to rewrite with `reduce()` (just prepend with prefix outside of the stream). Well, it's up to you.